File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] arr = new int[2];
        int n = nums.length;
        int start = 0;
        int end = n - 1;

        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (target > nums[mid]) {
                start = mid + 1;
            } else if (target < nums[mid]) {
                end = mid - 1;
            } else {
                arr[0] = findStart(nums, target, start, mid);
                arr[1] = findEnd(nums, target, mid, end);
                return arr;
            }
        }
        return new int[] { -1, -1 };

    }

    int findStart(int[] nums, int target, int start, int end) {
        while (start < end) {
            int mid = start + (end - start) / 2;
            if (nums[mid] == target) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }
        return start;
    }

    int findEnd(int[] nums, int target, int start, int end) {
        while (start < end) {
            int mid = start + (end - start) / 2 + 1; 
            if (nums[mid] == target) {
                start = mid;
            } else {
                end = mid - 1;
            }
        }
        return end;
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    let first = findFirstOccurence(nums, target);
    let last = lastOccurence(nums, target);

    return [first, last]
};


function findFirstOccurence(nums, target) {
    let n = nums.length;
    let low = 0;
    let high = n - 1;
    let first = -1;

    while(low <= high) {
        let mid = Math.floor((low + high) / 2);

        if(nums[mid] == target) {
            first = mid;
            high = mid - 1;
        } else if(nums[mid] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return first;
}

function lastOccurence(nums, target) {
   let n = nums.length;
    let low = 0;
    let high = n - 1;
    let last = -1;

    while(low <= high) {
        let mid = Math.floor((low + high) / 2);

        if(nums[mid] == target) {
            last = mid;
            low = mid + 1;
        } else if(nums[mid] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return last;
}